### PR TITLE
Apply macro expansion based on MutateInOpts

### PR DESCRIPTION
**Pull request by adamcfraser**
Uptake change in sg-bucket to provide the required macro expansions in a MutateInOpts parameter for the relevant methods, instead of hardcoding them into the implementations.

Currently the sg-bucket API only supports a single xattr per operation - will be expanded to multiple xattrs in a subsequent enhancement.

### CODE REVIEW

File: collection+subdoc.go [hunk 1]
```diff
@@ -145,6 +145,30 @@ func evalSubdocPath(subdoc any, path []string) (any, error) {
 	return subdoc, nil
 }
 
+// Upserts the value at the specified JSON path in the source.
+// Returns ErrPathMismatch if non-leaf path entries do not exist.
+func upsertSubdocValue(source any, path []string, value interface{}) error {
+
+	// eval path exists
+	subdoc, err := evalSubdocPath(source, path[0:len(path)-1])
```
Comment (torcolvin) [inline collection+subdoc.go:153]: readability nit:

I'd name `numLeaves := len(path) -1` because we use it again.

File: collection+subdoc.go [hunk 2]
```diff
@@ -145,6 +145,30 @@ func evalSubdocPath(subdoc any, path []string) (any, error) {
...
+	if err != nil {
+		return err
+	}
+
+	parent, ok := subdoc.(map[string]any)
```
Comment (torcolvin) [inline collection+subdoc.go:158]: Can we map this a compile time error by requiring map[string]any in compiler string?

File: collection+xattrs.go [hunk 3]
```diff
@@ -618,16 +567,63 @@ func removeUserXattrs(xattrs semiParsedXattrs) {
...
+		return e.casAsString(), nil
+	case sgbucket.MacroCrc32c:
+		return encodedCRC32c(e.value), nil
+	default:
+		return "", fmt.Errorf("Unsupported MacroExpansionType: %v", expansion)
+	}
+}
+
+// casAsString generates the string representation of CAS used by Couchbase Server's macro expansion
+func casAsString(value CAS) string {
```
Comment (torcolvin) [inline collection+xattrs.go:584]: nit: should this use `fmt.Stringer` ?

File: collection+xattrs.go [hunk 4]
```diff
@@ -618,16 +567,63 @@ func removeUserXattrs(xattrs semiParsedXattrs) {
...
+}
+
+// expandXattrMacros executes any macro expansions defined in mutateOpts
+func (e *event) expandXattrMacros(xattrKey string, xattr any, mutateOpts *sgbucket.MutateInOptions) error {
+
+	if mutateOpts == nil || len(mutateOpts.MacroExpansion) == 0 {
+		return nil
+	}
+
+	xattrMap, ok := xattr.(map[string]any)
```
Comment (torcolvin) [inline collection+xattrs.go:601]: can we make this a compile time error? and check for nil here?

File: collection+xattrs.go [hunk 5]
```diff
@@ -618,16 +567,63 @@ func removeUserXattrs(xattrs semiParsedXattrs) {
...
-		xattrMap["value_crc32c"] = encodedCRC32c(e.value)
+	// loop through provided macro expansions, check they apply to this xattr, then upsert into the xattr
+	// at the specified path
+	for _, v := range mutateOpts.MacroExpansion {
+		path, err := parseSubdocPath(v.Path)
+		if err != nil {
+			return err
+		}
+		if path[0] != xattrKey {
+			return fmt.Errorf("Unknown xattr in macro expansion path: %s", path)
```
Comment (torcolvin) [inline collection+xattrs.go:614]: Should we list path[0] in this error, and maybe the xattr we expect?

File: collection+xattrs.go [hunk 6]
```diff
@@ -618,16 +567,63 @@ func removeUserXattrs(xattrs semiParsedXattrs) {
...
+		}
+
+		expandedValue, err := e.macroExpand(v.Type)
+		if err != nil {
+			return err
+		}
+
+		if err := upsertSubdocValue(xattrMap, path[1:], expandedValue); err != nil {
+			return fmt.Errorf("Unable to set macro expansion value at path: %v: %v", v.Path, err)
```
Comment (torcolvin) [inline collection+xattrs.go:623]: should this be path[1:] that is logged here?

Can you wrap the error here?

**Review by torcolvin — COMMENTED** [review]
These are largely nitpick comments for test code, but might help us debug things in the future. Feel free to disregard.

File: collection+subdoc.go (hunk 1 above)
Comment (adamcfraser) [reply collection+subdoc.go:153]: I evaluated this, but prefer leaving as len(path)-1 as I feel `lastPath := path[len(path)-1]` is more readable than having to go back to see what numLeaves is.

File: collection+subdoc.go (hunk 2 above)
Comment (adamcfraser) [reply collection+subdoc.go:158]: This is required to support non-map JSON.

File: collection+xattrs.go (hunk 3 above)
Comment (adamcfraser) [reply collection+xattrs.go:584]: Intentionally avoided this as I think we'd prefer to report CAS normally when logging, and only use this format for compatibility with CBS.

File: collection+xattrs.go (hunk 4 above)
Comment (adamcfraser) [reply collection+xattrs.go:601]: As above, required for generic JSON support.

File: collection+xattrs.go (hunk 5 above)
Comment (adamcfraser) [reply collection+xattrs.go:614]: Agree - fixed.

File: collection+xattrs.go (hunk 6 above)
Comment (adamcfraser) [reply collection+xattrs.go:623]: I think full path is preferable for context, but agree on wrapping the error.

**Review by torcolvin — APPROVED** [review]
LGTM, will need to bump sg-bucket once the merge happens.

**Review by torcolvin — APPROVED** [review]

**Pull request merged**